package/base-files/files/lib/functions/system.sh

```diff
@@ -145,91 +156,131 @@ mmc_get_mac_binary() {
 }
 
 macaddr_add() {
-	local mac=$1
-	local val=$2
-	local oui=${mac%:*:*:*}
-	local nic=${mac#*:*:*:}
+	local mac="$1"
+	local val="${2:-1}"
+	local oui nic
 
-	nic=$(printf "%06x" $((0x${nic//:/} + val & 0xffffff)) | sed 's/^\(.\{2\}\)\(.\{2\}\)\(.\{2\}\)/\1:\2:\3/')
-	echo $oui:$nic
+	oui=$(macaddr_octet "$mac" 0)
+	nic=$(macaddr_octet "$mac" 3)
+	printf '%s%06x' "$oui" $(((0x${nic} + val) & 0xffffff)) | macaddr_canonicalize
 }
 
-macaddr_geteui() {
-	local mac=$1
-	local sep=$2
+macaddr_octet() {
+	local mac="$1"
+	local off="${2:-3}"
+	local len="${3:-3}"
+	local sep="$4"
+	local oct=""
+
+	while [ "$off" -ne 0 ]; do
+		mac="${mac#*$delimiter}"
+		off=$((off - 1))
+	done
+
+	while [ "$len" -ne 0 ]; do
+		oct="${oct}${oct:+$sep}${mac%%$delimiter*}"
+		mac="${mac#*$delimiter}"
+		len=$((len - 1))
+	done
 
-	echo ${mac:9:2}$sep${mac:12:2}$sep${mac:15:2}
+	echo -n "$oct"
 }
 
 macaddr_setbit() {
-	local mac=$1
-	local bit=${2:-0}
+	local hex="${1//$delimiter/}"
+	local bit="${2:-0}"
```

Comment:
The reason why I set this to 0 as default in dc61e3b7ffdc was because I wanted to make sure the function returns nothing if `bit` argument is not defined. If you're going to make it start from 0 (which I don't like), then I think the default value should be changed too. (why prefer certain bit over others?)